transforms/ExplicitConsdes.cpp

```diff
@@ -0,0 +1,189 @@
+
+            FunctionEntry entry{};
+            entry.priority =
+                dyn_cast<ConstantInt>(elem->getAggregateElement(0u))
+                    ->getZExtValue();
+            entry.function = elem->getAggregateElement(1);
+            entry.param = elem->getAggregateElement(2);
+            target.push_back(std::move(entry));
+        }
+    }
+
+    void insertCalls(std::vector<FunctionEntry> &entries, Instruction *before) {
+        Instruction *after = nullptr;
+
+        for (auto &entry : entries) {
+            auto *call = CallInst::Create(
+                dyn_cast<FunctionType>(
+                    entry.function->getType()->getPointerElementType()),
+                entry.function, "");
+            call->setDebugLoc(DebugLoc::get(0,0, before->getDebugLoc().getScope()));
+            if (after) {
+                call->insertAfter(after);
+            } else {
+                call->insertBefore(before);
+            }
+            after = call;
+        }
+    }
+
+   public:
+    static char ID;
+
+    ExplicitConsdes() : ModulePass(ID) {}
+
+    bool runOnModule(Module &mod) override {
+        using std::sort;
+        using std::tuple;
+        using std::vector;
+
+        bool change = false;
+
+        vector<FunctionEntry> ctors, dtors;
+
+        GlobalVariable *dtorsVar = nullptr;
+        GlobalVariable *ctorsVar = nullptr;
+
+        for (auto &glob : mod.globals()) {
+            if (glob.hasName() && glob.getName() == "llvm.global_dtors") {
+                fillFromInitializer(dtors, glob.getInitializer());
+                dtorsVar = &glob;
+            }
+
+            if (glob.hasName() && glob.getName() == "llvm.global_ctors" &&
+                glob.hasInitializer()) {
+                fillFromInitializer(ctors, glob.getInitializer());
+                ctorsVar = &glob;
+            }
+        }
+
+        if (ctors.empty() && dtors.empty()) {
+            return false;
+        }
+
+        // sort them to call order
+        sort(ctors.begin(), ctors.end(), priorityAsc);
+        sort(dtors.begin(), dtors.end(), priorityDesc);
+
+        // call constructors at the beginning of main
+        auto *main = mod.getFunction("main");
+        auto *ctorsBefore =
+            main->getEntryBlock().getFirstNonPHIOrDbgOrLifetime();
+        insertCalls(ctors, ctorsBefore);
+
+        vector<Instruction *> dtorsBefore;
+
+        // search for returns in main
+        for (auto &block : *main) {
+            if (isa<ReturnInst>(block.getTerminator())) {
+                dtorsBefore.push_back(block.getTerminator());
+            }
+        }
+
+        // and finally search for calls to exit in whole program
+        for (auto &func : mod.functions()) {
+            for (auto &block : func) {
+                for (auto &inst : block) {
+                    if (auto *call = dyn_cast<CallInst>(&inst)) {
+                        if (call->isIndirectCall()) continue;
```

Comment:
Hmm, yes, that is the best we can do here... Unless we want to plug in pointer analysis. Let's hope nobody calls `exit` via pointers :)

Reply:
Yeah, the only solution applicable here I can think of is to output a warning that a pointer to `exit` is being stored somewhere. I thought we have the same limitation while instrumenting memsafety. Would we even detect indirect call of malloc?

Reply:
Yes, that is probably a good idea to give at least a warning.

Reply:
> Would we even detect indirect call of malloc
I'm not sure. I know that slicer had support for that in the past, but it probably does not work right now. And instrumentation? Probably no.